core/src/main/scala/kafka/log/ProducerStateManager.scala

```diff
@@ -48,33 +48,80 @@ private[log] case class TxnMetadata(producerId: Long, var firstOffset: LogOffset
-    -1, 0, RecordBatch.NO_TIMESTAMP, -1, None)
+  private[log] val NumBatchesToRetain = 5
+  def empty(producerId: Long) = new ProducerIdEntry(producerId, mutable.Queue[BatchMetadata](), RecordBatch.NO_PRODUCER_EPOCH, -1, None)
 }
 
-private[log] case class ProducerIdEntry(producerId: Long, producerEpoch: Short, lastSeq: Int, lastOffset: Long,
-                                        offsetDelta: Int, timestamp: Long, coordinatorEpoch: Int,
-                                        currentTxnFirstOffset: Option[Long]) {
-  def firstSeq: Int = lastSeq - offsetDelta
-  def firstOffset: Long = lastOffset - offsetDelta
+private[log] case class BatchMetadata(lastSeq: Int, lastOffset: Long, offsetDelta: Int, timestamp: Long) {
+  def firstSeq = lastSeq - offsetDelta
+  def firstOffset = lastOffset - offsetDelta
 
-  def isDuplicate(batch: RecordBatch): Boolean = {
-    batch.producerEpoch == producerEpoch &&
-      batch.baseSequence == firstSeq &&
-      batch.lastSequence == lastSeq
+  override def toString: String = {
+    "BatchMetadata(" +
+      s"firstSeq=$firstSeq, " +
+      s"lastSeq=$lastSeq, " +
+      s"firstOffset=$firstOffset, " +
+      s"lastOffset=$lastOffset, " +
+      s"timestamp=$timestamp)"
+  }
+}
+
+// the batchMetadata is ordered such that the batch with the lowest sequence is at the head of the queue while the
+// batch with the highest sequence is at the tail of the queue. We will retain at most ProducerIdEntry.NumBatchesToRetain
+// elements in the queue. When the queue is at capacity, we remove the first element to make space for the incoming batch.
+private[log] class ProducerIdEntry(val producerId: Long, val batchMetadata: mutable.Queue[BatchMetadata],
+                                   var producerEpoch: Short, var coordinatorEpoch: Int,
+                                   var currentTxnFirstOffset: Option[Long]) {
+
+  def firstSeq: Int = if (batchMetadata.isEmpty) RecordBatch.NO_SEQUENCE else batchMetadata.front.firstSeq
+  def firstOffset: Long = if (batchMetadata.isEmpty) -1L else batchMetadata.front.firstOffset
+
+  def lastSeq: Int = if (batchMetadata.isEmpty) RecordBatch.NO_SEQUENCE else batchMetadata.last.lastSeq
+  def lastOffset: Long = if (batchMetadata.isEmpty) -1L else batchMetadata.last.lastOffset
+  def lastTimestamp = if (batchMetadata.isEmpty) RecordBatch.NO_TIMESTAMP else batchMetadata.last.timestamp
+
+  def addBatchMetadata(producerEpoch: Short, lastSeq: Int, lastOffset: Long, offsetDelta: Int, timestamp: Long) = {
+    maybeUpdateEpoch(producerEpoch)
+
+    if (batchMetadata.size == ProducerIdEntry.NumBatchesToRetain)
+      batchMetadata.dequeue()
+
+    batchMetadata.enqueue(BatchMetadata(lastSeq, lastOffset, offsetDelta, timestamp))
+  }
+
+  def maybeUpdateEpoch(producerEpoch: Short): Boolean = {
+    if (this.producerEpoch != producerEpoch) {
+      batchMetadata.clear()
+      this.producerEpoch = producerEpoch
+      return true
```

Comment:
nit: explicit returns are frowned upon in scala. Maybe we can drop the `return` and add an `else` branch? Same for `duplicateOf` below.